tests/pyfunc/test_spark.py

```diff
@@ -33,44 +32,78 @@ def score_model_as_udf(model_path, run_id, pandas_df, result_type="double"):
-                             dtype=types[i]) for i in range(len(prediction))},
-            columns=[str(i) for i in range(len(prediction))])
-        return prediction_df
+def get_model_class():
+    """
+    Defines a custom Python model class that wraps a scikit-learn estimator.
+    This can be invoked within a pytest fixture to define the class in the ``__main__`` scope.
+    Alternatively, it can be invoked within a module to define the class in the module's scope.
+    """
+    class ConstantModel(mlflow.pyfunc.PythonModel):
+        def predict(self, model_input):
+            m, _ = model_input.shape
+            prediction_df = pd.DataFrame(data={
+                str(i): np.array([prediction[i] for j in range(m)],
+                                 dtype=types[i]) for i in range(len(prediction))},
+                columns=[str(i) for i in range(len(prediction))])
+            return prediction_df
+
+    return ConstantModel
+
+
+class ModuleScopedConstantModel(get_model_class()):
+    """
+    A custom Python model class defined in the test module scope.
+    """
+    pass
 
 
 def _load_pyfunc(_):
-    return ConstPyfunc
-
-
-class TestSparkUDFs(unittest.TestCase):
-    def setUp(self):
-        os.environ["PYSPARK_PYTHON"] = sys.executable
-        self._tmp = tempfile.mkdtemp("mlflow-spark-test", dir="/tmp")
-        # NB: local-cluster mode actually sets up 2 executors, each given 1 core
-        # and 1024 MB of memory. This is the best way to simulate pickling/serialization
-        # behavior to ensure it will work as expected on a real cluster.
-        self.spark = pyspark.sql.SparkSession.builder \
-            .config(key="spark.python.worker.reuse", value=True) \
-            .master("local-cluster[2, 1, 1024]") \
-            .getOrCreate()
-        self._model_path = os.path.join(self._tmp, "model")
-        mlflow.pyfunc.save_model(self._model_path,
-                                 loader_module=os.path.basename(__file__)[:-3],
-                                 code_path=[__file__])
-
-    def tearDown(self):
-        shutil.rmtree(self._tmp)
-
-    @pytest.mark.large
-    def test_spark_udf(self):
+    return ModuleScopedConstantModel()
+
+
+@pytest.fixture
+def main_scoped_model_class():
+    """
+    A custom Python model class defined in the ``__main__`` scope.
+    """
+    return get_model_class()
+
+
+@pytest.fixture(autouse=True)
+def configure_environment():
+    os.environ["PYSPARK_PYTHON"] = sys.executable
+
+
+@pytest.fixture
+def spark():
+    return pyspark.sql.SparkSession.builder\
+        .config(key="spark.python.worker.reuse", value=True)\
+        .master("local-cluster[2, 1, 1024]")\
+        .getOrCreate()
+
+
+@pytest.fixture
+def model_path(tmpdir):
+    return os.path.join(str(tmpdir), "model")
+
+
+@pytest.mark.large
+def test_spark_udf(spark, main_scoped_model_class, tmpdir):
+    class_instance_model_path = os.path.join(str(tmpdir), "class_model")
+    mlflow.pyfunc.save_model(
```

Comment:
Maybe comment at the top of this test that we plan to save the model in two different ways (although ideally these are two separate tests which just share code; if this test fails for one of the two branches, it may be challenging to figure out which one failed).

Reply:
I also don't love that you felt the need to add an integration test here; ideally, the contract and architecture of pyfunc save_model should ensure that it's not possible that one works and the other doesn't. This might be an argument for implementing save_model with loader_module in terms of PythonModel.